### PR TITLE
Created new modules for the future constructor functionality (Feature/569)

### PR DESCRIPTION
**Pull request by twihell**
## Description

1. Created new modules for additional constructor options and restrictions:

- pockets
- backs
- straps
- restrictions

2. Refactored patterns, closures, constructorBasic, constructorBottom, constructorFrontPocket, and model modules and added new fields into them.

3. Created Joi validation for all constructor-related options and implemented them in permissions

4. Added history logging for all constructor-related modules

5. Created migrations that need to be applied to the remote database (not all of them are automatic, i.e. for some collections I will need to make individual document insertions step-by-step)

6. Fixed tests that failed after my changes

### Checklist
- [x] 🔽 My branch is up-to-date with "development" branch
- [x] ✅All tests passed locally
- [x] ✨My changes working with up-to-date front-end and admin part locally, like charm
- [x] 🔗 Link pull request to issue

### CODE REVIEW

File: modules/back/back.service.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    const foundBack = await Back.findById(id).exec();
+    if (foundBack) {
+      return foundBack;
+    }
+    throw new RuleError(BACK_NOT_FOUND, NOT_FOUND);
+  }
+
+  async getBacksByModel(id) {
+    const back = Back.find({ model: id }).exec();
+    if (back.length === 0) {
```
Comment (AndriyLymych) [inline modules/back/back.service.js:64]: use if (!back)

File: modules/back/back.service.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    return back;
+  }
+
+  async updateBack(id, back, image, { _id: adminId }) {
+    const backToUpdate = await Back.findById(id).exec();
+    if (!backToUpdate) {
+      throw new RuleError(BACK_NOT_FOUND, NOT_FOUND);
+    }
+
+    const existChecker = await Promise.resolve(this.checkIfBackExist(back, id));
```
Comment (AndriyLymych) [inline modules/back/back.service.js:76]: it will be better create middleware for this action

File: modules/back/back.service.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      );
+
+      await addHistoryRecord(historyRecord);
+
+      return Back.findByIdAndDelete(id);
+    }
+  }
+
+  async addBack(back, image, { _id: adminId }) {
+    const existChecker = await Promise.resolve(this.checkIfBackExist(back));
```
Comment (AndriyLymych) [inline modules/back/back.service.js:147]: same issue

File: modules/back/back.service.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      name: {
+        $elemMatch: {
+          $or: data.name.map(({ value }) => ({ value })),
+        },
+      },
+    }).exec();
+
+    return backCount > 0;
+  }
+}
```
Comment (AndriyLymych) [inline modules/back/back.service.js:197]: put it func into middleware

File: modules/closures/closures.graphql.js [hunk 5]
```diff
@@ -2,22 +2,43 @@ const closureType = `
...
     _id: ID
     name: [Language]
-    material: Material
+    optionType: String
+    model: Model
+    features: ClosureFeatureSet
     image: String
     additionalPrice: [CurrencySet]
     available: Boolean
+    default: Boolean
```
Comment (AndriyLymych) [inline modules/closures/closures.graphql.js:11]: use correct naming

File: modules/closures/closures.graphql.js [hunk 6]
```diff
@@ -2,22 +2,43 @@ const closureType = `
...
     name: [LanguageInput]
-    material: ID
+    optionType: String
+    model: ID
+    features: ClosureFeatureSetInput
     image: Upload
-    additionalPrice: [CurrencySetInput]
+    additionalPrice: Int
     available: Boolean
+    default: Boolean
```
Comment (AndriyLymych) [inline modules/closures/closures.graphql.js:31]: same

File: modules/closures/closures.model.js [hunk 7]
```diff
@@ -3,18 +3,30 @@ const mongoose = require('mongoose');
...
+    },
+    color: {
+      type: mongoose.Schema.Types.ObjectId,
+      ref: COLOR,
+    },
   },
   image: String,
   additionalPrice: [CurrencySet],
   available: Boolean,
+  default: Boolean,
```
Comment (AndriyLymych) [inline modules/closures/closures.model.js:29]: same

File: modules/back/back.service.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    AVAILABLE,
+    FEATURES,
+    OPTION_TYPE,
+    MODEL,
+  },
+} = require('../../consts/history-obj-keys');
+
+class BackService {
+  async getAllBacks({ skip, limit }) {
+    const items = await Back.find()
```
Comment (AndriyLymych) [inline modules/back/back.service.js:40]: you need filters for this

File: modules/constructor/constructor-bottom/constructor-bottom.graphql.js [hunk 9]
```diff
@@ -2,28 +2,43 @@ const constructorBottomType = `
...
+    optionType: String
+    model: ID
+    features: ConstructorBottomFeatureInput
     image: String
-    basePrice: Int!
-    available: Boolean!
+    basePrice: Int
+    available: Boolean
     default:Boolean
   }
```
Comment (AndriyLymych) [inline modules/constructor/constructor-bottom/constructor-bottom.graphql.js:32]: same

File: modules/constructor/constructor-front-pocket/constructor-front-pocket.graphgl.js [hunk 10]
```diff
@@ -1,29 +1,46 @@
...
+    name: [Language]
+    optionType: String
+    model: Model
+    features: ConstructorFrPocketFeatureSet
+    image:String
     basePrice: [CurrencySet]
-    available: Boolean,
-    default:Boolean,
+    available: Boolean
+    default:Boolean
```
Comment (AndriyLymych) [inline modules/constructor/constructor-front-pocket/constructor-front-pocket.graphgl.js:11]: same

File: modules/constructor/constructor-front-pocket/constructor-front-pocket.graphgl.js [hunk 11]
```diff
@@ -1,29 +1,46 @@
...
+const constructorFrontPocketInputs = `  
+input ConstructorFrontPocketInput {
+  name: [LanguageInput]
+  optionType: String
+  model: ID
+  features: ConstructorFrPocketFeatureInput
+  image: String
+  basePrice: Int
+  available: Boolean
+  default:Boolean
```
Comment (AndriyLymych) [inline modules/constructor/constructor-front-pocket/constructor-front-pocket.graphgl.js:32]: same

File: modules/constructor/constructor.services.js [hunk 12]
```diff
@@ -1,51 +1,115 @@
...
+  ) {
     if (upload) {
       constructorElement.image = await uploadSmallImage(upload);
     }
-    if (await this.checkConstructorElementExist(constructorElement, model)) {
-      throw new Error(error);
+
+    const constrElemChecker = await Promise.resolve(
+      this.checkConstructorElementExist(constructorElement, model)
+    );
```
Comment (AndriyLymych) [inline modules/constructor/constructor.services.js:72]: should be better put checkConstructorElementExist into middlewares

File: modules/model/model.resolver.js [hunk 13]
```diff
@@ -35,9 +35,9 @@ const modelsQuery = {
 };
 
 const modelsMutation = {
-  addModel: async (parent, args) => {
+  addModel: async (parent, args, { user }) => {
     try {
-      return await modelsService.addModel(args.model, args.upload);
+      return await modelsService.addModel(args.model, args.upload, user);
     } catch (e) {
       return {
```
Comment (AndriyLymych) [inline modules/model/model.resolver.js:42]: replace with Rule Error

File: modules/model/model.resolver.js [hunk 14]
```diff
@@ -46,9 +46,14 @@ const modelsMutation = {
...
     try {
-      return await modelService.updateModel(args.id, args.model, args.upload);
+      return await modelService.updateModel(
+        args.id,
+        args.model,
+        args.upload,
+        user
+      );
     } catch (e) {
       return {
```
Comment (AndriyLymych) [inline modules/model/model.resolver.js:58]: same

File: modules/model/model.resolver.js [hunk 15]
```diff
@@ -57,9 +62,9 @@ const modelsMutation = {
...
   },
 
-  deleteModel: async (parent, args) => {
+  deleteModel: async (parent, args, { user }) => {
     try {
-      return await modelsService.deleteModel(args.id);
+      return await modelsService.deleteModel(args.id, user);
     } catch (e) {
       return {
         statusCode: NOT_FOUND,
```
Comment (AndriyLymych) [inline modules/model/model.resolver.js:70]: same

File: modules/model/model.service.js [hunk 16]
```diff
@@ -51,7 +78,7 @@ class ModelsService {
     return Model.find({ category: id });
   }
 
-  async addModel(data, upload) {
+  async addModel(data, upload, { _id: adminId }) {
     if (await this.checkModelExist(data)) {
```
Comment (AndriyLymych) [inline modules/model/model.service.js:82]: its not cool doing call to db in if statemant

File: modules/pattern/pattern.graphql.js [hunk 17]
```diff
@@ -2,25 +2,47 @@ const patternType = `
...
       description: [LanguageInput]
+      optionType: String
+      model: ID
+      features: PatternFeatureSetInput
       images:ImageSetInput
-      constructorImg: String,
+      constructorImg: String
+      additionalPrice: Int
+      available: Boolean
+      default: Boolean
```
Comment (AndriyLymych) [inline modules/pattern/pattern.graphql.js:35]: same

File: modules/pattern/pattern.model.js [hunk 18]
```diff
@@ -1,21 +1,31 @@
...
+      ref: MATERIAL,
+    },
+    handmade: Boolean,
   },
-  handmade: Boolean,
+  images: ImageSet,
+  constructorImg: String,
+  additionalPrice: [CurrencySet],
   available: Boolean,
+  default: Boolean,
```
Comment (AndriyLymych) [inline modules/pattern/pattern.model.js:28]: same

File: modules/pattern/pattern.service.js [hunk 19]
```diff
@@ -19,7 +24,15 @@ const {
...
+    DESCRIPTION,
+    AVAILABLE,
+    FEATURES,
+    ADDITIONAL_PRICE,
+    MODEL,
+    OPTION_TYPE,
+  },
 } = require('../../consts/history-obj-keys');
 
 class PatternsService {
```
Comment (AndriyLymych) [inline modules/pattern/pattern.service.js:38]: need add filters for patterns

File: modules/pattern/pattern.service.js [hunk 20]
```diff
@@ -43,24 +56,34 @@ class PatternsService {
...
     const patternToUpdate = await Pattern.findById(id).exec();
     if (!patternToUpdate) {
-      throw new Error(PATTERN_NOT_FOUND);
+      throw new RuleError(PATTERN_NOT_FOUND, NOT_FOUND);
     }
-    if (await this.checkPatternExist(pattern, id)) {
-      throw new Error(PATTERN_ALREADY_EXIST);
+
+    const existChecker = await Promise.resolve(
+      this.checkPatternExist(pattern, id)
```
Comment (AndriyLymych) [inline modules/pattern/pattern.service.js:69]: should be better put it into middlewares

File: modules/pattern/pattern.service.js [hunk 21]
```diff
@@ -99,33 +123,42 @@ class PatternsService {
   }
 
   async addPattern({ pattern, image }, { _id: adminId }) {
-    if (await this.checkPatternExist(pattern)) {
-      throw new Error(PATTERN_ALREADY_EXIST);
+    const existChecker = await Promise.resolve(this.checkPatternExist(pattern));
```
Comment (AndriyLymych) [inline modules/pattern/pattern.service.js:126]: same

File: modules/pattern/pattern.service.js [hunk 22]
```diff
@@ -99,33 +123,42 @@ class PatternsService {
...
+
+    if (existChecker) {
+      throw new RuleError(PATTERN_ALREADY_EXIST, BAD_REQUEST);
     }
 
-    if (image.length) {
-      if (!image[0] && !image[1]) {
+    if (await image.length) {
+      if ((await !image[0]) && !image[1]) {
```
Comment (AndriyLymych) [inline modules/pattern/pattern.service.js:133]: it is not correctly

File: modules/pocket/pocket.graphql.js [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+  type Pocket {
+    _id: ID!
+    name: [Language]
+    optionType: String
+    model: Model
+    features: [PocketSide]
+    image: String
+    additionalPrice: [CurrencySet]
+    available: Boolean
+    default: Boolean
```
Comment (AndriyLymych) [inline modules/pocket/pocket.graphql.js:11]: same

File: modules/pocket/pocket.graphql.js [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+const pocketInput = `
+  input PocketInput {
+    name: [LanguageInput]
+    optionType: String
+    model: ID
+    features: PocketSideInput
+    image: Upload
+    additionalPrice: Int
+    available: Boolean
+    default: Boolean
```
Comment (AndriyLymych) [inline modules/pocket/pocket.graphql.js:45]: same

File: modules/pocket/pocket.model.js [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+        type: String,
+        enum: [RIGHT, LEFT, FRONT, BACK],
+        required: true,
+      },
+    ],
+  },
+  image: String,
+  additionalPrice: [CurrencySet],
+  available: Boolean,
+  default: Boolean,
```
Comment (AndriyLymych) [inline modules/pocket/pocket.model.js:29]: same

File: modules/pocket/pocket.resolver.js [hunk 26]
```diff
@@ -0,0 +1,52 @@
+const pocketService = require('./pocket.service');
+
+const RuleError = require('../../errors/rule.error');
+
+const pocketQuery = {
+  getAllPockets: (parent, args) => pocketService.getAllPockets(args),
```
Comment (AndriyLymych) [inline modules/pocket/pocket.resolver.js:6]: do try...catch for this

File: modules/pocket/pocket.service.js [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+    ADDITIONAL_PRICE,
+    AVAILABLE,
+    FEATURES,
+    OPTION_TYPE,
+    MODEL,
+  },
+} = require('../../consts/history-obj-keys');
+
+class PocketService {
+  async getAllPockets({ skip, limit }) {
```
Comment (AndriyLymych) [inline modules/pocket/pocket.service.js:36]: need filters for this case

File: modules/pocket/pocket.service.js [hunk 28]
```diff
@@ -0,0 +1,197 @@
...
+    const foundPocket = await Pocket.findById(id).exec();
+    if (foundPocket) {
+      return foundPocket;
+    }
+    throw new RuleError(POCKET_NOT_FOUND, NOT_FOUND);
+  }
+
+  async getPocketsByModel(id) {
+    const pocket = Pocket.find({ model: id }).exec();
+    if (pocket.length === 0) {
```
Comment (AndriyLymych) [inline modules/pocket/pocket.service.js:61]: same

File: modules/pocket/pocket.service.js [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+    let pocketCount = await Pocket.countDocuments({
+      _id: { $ne: id },
+      name: {
+        $elemMatch: {
+          $or: data.name.map(({ value }) => ({ value })),
+        },
+      },
+    }).exec();
+    return pocketCount > 0;
+  }
```
Comment (AndriyLymych) [inline modules/pocket/pocket.service.js:194]: put it into middleware

File: modules/restriction/restriction.resolver.js [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+      return await restrictionService.deleteRestriction(args.id, user);
+    } catch (e) {
+      return new RuleError(e.message, e.statusCode);
+    }
+  },
+};
+
+const restrictionQuery = {
+  getAllRestrictions: (parent, args) =>
+    restrictionService.getAllRestrictions(args),
```
Comment (AndriyLymych) [inline modules/restriction/restriction.resolver.js:34]: same

File: modules/restriction/restriction.service.js [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+  getChanges,
+  generateHistoryChangesData,
+} = require('../../utils/hisrory');
+const { addHistoryRecord } = require('../history/history.service');
+const {
+  HISTORY_OBJ_KEYS: { COMPARE_BY_EXPRESSION, OPTIONS },
+} = require('../../consts/history-obj-keys');
+
+class RestrictionService {
+  async getAllRestrictions({ skip, limit }) {
```
Comment (AndriyLymych) [inline modules/restriction/restriction.service.js:23]: need filters

File: modules/strap/strap.graphql.js [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+const strapInputs = `
+  input StrapInput {
+    name: [LanguageInput]
+    optionType: String
+    model: ID
+    features: StrapFeatureInput
+    image: Upload
+    additionalPrice: Int
+    available: Boolean
+    default: Boolean
```
Comment (AndriyLymych) [inline modules/strap/strap.graphql.js:31]: same

File: modules/strap/strap.model.js [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+  features: {
+    color: {
+      type: mongoose.Schema.Types.ObjectId,
+      ref: COLOR,
+    },
+  },
+  image: String,
+  additionalPrice: [CurrencySet],
+  available: Boolean,
+  default: Boolean,
```
Comment (AndriyLymych) [inline modules/strap/strap.model.js:25]: same

File: modules/strap/strap.resolver.js [hunk 34]
```diff
@@ -0,0 +1,52 @@
+const strapService = require('./strap.service');
+
+const RuleError = require('../../errors/rule.error');
+
+const strapQuery = {
+  getAllStraps: (parent, args) => strapService.getAllStraps(args),
```
Comment (AndriyLymych) [inline modules/strap/strap.resolver.js:6]: same

File: modules/strap/strap.service.js [hunk 35]
```diff
@@ -0,0 +1,196 @@
...
+    ADDITIONAL_PRICE,
+    AVAILABLE,
+    FEATURES,
+    OPTION_TYPE,
+    MODEL,
+  },
+} = require('../../consts/history-obj-keys');
+
+class StrapService {
+  async getAllStraps({ skip, limit }) {
```
Comment (AndriyLymych) [inline modules/strap/strap.service.js:36]: need filters

File: modules/strap/strap.service.js [hunk 36]
```diff
@@ -0,0 +1,196 @@
...
+    const foundStrap = await Strap.findById(id).exec();
+    if (foundStrap) {
+      return foundStrap;
+    }
+    throw new RuleError(STRAP_NOT_FOUND, NOT_FOUND);
+  }
+
+  async getStrapsByModel(id) {
+    const strap = Strap.find({ model: id }).exec();
+    if (strap.length === 0) {
```
Comment (AndriyLymych) [inline modules/strap/strap.service.js:61]: same

File: modules/strap/strap.service.js [hunk 37]
```diff
@@ -0,0 +1,196 @@
...
+      _id: { $ne: id },
+      name: {
+        $elemMatch: {
+          $or: data.name.map(({ value }) => ({ value })),
+        },
+      },
+    }).exec();
+    return strapCount > 0;
+  }
+}
```
Comment (AndriyLymych) [inline modules/strap/strap.service.js:194]: same

File: validators/constructor-items-inputs.validator.js [hunk 38]
```diff
@@ -0,0 +1,218 @@
...
+const nestedNameValidator = Joi.object({
+  lang: Joi.string()
+    .trim()
+    .required(),
+  value: Joi.string()
+    .trim()
+    .required(),
+});
+
+const crudPocketValidator = Joi.object({
```
Comment (AndriyLymych) [inline validators/constructor-items-inputs.validator.js:27]: use just pocketValidator

File: validators/constructor-items-inputs.validator.js [hunk 39]
```diff
@@ -0,0 +1,218 @@
...
+module.exports = {
+  crudPocketValidator,
+  crudBackValidator,
+  crudClosureValidator,
+  crudStrapValidator,
+  crudPatternValidator,
+  crudConstructorBasicValidator,
+  crudConstructorBottomValidator,
+  crudConstrFrontPocketValidator,
+};
```
Comment (AndriyLymych) [inline validators/constructor-items-inputs.validator.js:218]: dont doing  duplication

File: validators/constructor-items.validator.js [hunk 40]
```diff
@@ -0,0 +1,253 @@
...
+  pocketValidator,
+  backValidator,
+  closureValidator,
+  strapValidator,
+  patternValidator,
+  constructorBasicValidator,
+  constructorBottomValidator,
+  constructorFrontPocketValidator,
+  restrictionValidator,
+};
```
Comment (AndriyLymych) [inline validators/constructor-items.validator.js:253]: same

**Comment by yuliiakucher** [comment]
Please fix duplication
image

File: .vscode/settings.json [hunk 41]
```diff
@@ -1,3 +0,0 @@
-{
-    "liveServer.settings.port": 5501
```
Comment (yuliiakucher) [inline .vscode/settings.json:2]: please dont change this

File: modules/back/back.resolver.js [hunk 42]
```diff
@@ -0,0 +1,50 @@
+const backService = require('./back.service');
+
+const RuleError = require('../../errors/rule.error');
+
+const backQuery = {
+  getAllBacks: (parent, args) => backService.getAllBacks(args),
```
Comment (AndriyLymych) [inline modules/back/back.resolver.js:6]: add try catch

File: modules/back/back.service.js [hunk 43]
```diff
@@ -0,0 +1,204 @@
...
+        backCount = calculations.count;
+      }
+
+      return {
+        items,
+        count: backCount || 0,
+      };
+    } catch (e) {
+      throw new RuleError(e.message, e.statusCode);
+    }
```
Comment (AndriyLymych) [inline modules/back/back.service.js:71]: you don't need try catch here

File: modules/back/back.graphql.js [hunk 44]
```diff
@@ -0,0 +1,44 @@
...
+    image: Upload
+    additionalPrice: Int
+    available: Boolean
+    customizable: Boolean
+  }
+
+  input BackFeatureSetInput {
+    material: ID
+    color: ID
+  }
```
Comment (AndriyLymych) [inline modules/back/back.graphql.js:37]: add filter input

File: modules/back/back.service.js [hunk 45]
```diff
@@ -0,0 +1,204 @@
...
+  }
+
+  async getBackById(id) {
+    const foundBack = await Back.findById(id).exec();
+    if (foundBack) {
+      return foundBack;
+    }
```
Comment (AndriyLymych) [inline modules/back/back.service.js:78]: fix to one style

**Review by AndriyLymych — APPROVED** [review]

**Review by yuliiakucher — APPROVED** [review]

**Review by grebiniuk — APPROVED** [review]
Thank you!

**Pull request merged**